python/run-tests.py

```diff
@@ -107,20 +118,26 @@ def run_individual_python_test(target_dir, test_name, pyspark_python):
     env["PYSPARK_SUBMIT_ARGS"] = " ".join(spark_args)
 
     output_prefix = get_valid_filename(pyspark_python + "__" + test_name + "__").lstrip("_")
-    per_test_output = tempfile.NamedTemporaryFile(prefix=output_prefix, suffix=".log")
+
+    if keep_test_output:
+        # The location is unique because the test is already in a unique directory.
```

Comment:
I still think we should better to expose an option as a directory path, and just don't remove them because with this option on, the output path becomes different anyway.

Reply:
I was actually looking into this first. However, if you look at the above code we already pass in a target directory that is used for the hive warehouse path. For debugging purposes it seems related that if you want to retain the test output the actually written files will be interesting as well. 

As for this, my proposal here would be to reuse the existing target dir option and simply move the log output into this path.

If we add another path option it becomes harder to reconcile why one is deleted and the other one is not and we would need an additional flag to to allow retaining the other one too.

WDYT?

Reply:
I believe we use `target/` directory so `mvn clean` or `sbt clean` remove them out. If we specify a directory (for the purpose of keeping), I think it makes sense to allow other directories. Maybe we could just document that it will not be automatically removed if we you specify the path.

Reply:
Ok, so what about we do the following:

* I will add an argument that allows us to override the target directory so that we harmonize the way it's used today and not implicitly create the `target` name.

The only thing I'm wondering though is from a developer feeling, you want to retain the output, but now you need to remember two things, one which is that the option is actually a path and you need to specify a path name.

Whereas in the other case, the developer needs to remember to just specify `-k` to keep the output, and if it's supposed to survive `mvn clean` use `--target myothertarget`. I personally find the flag easier to use but I will not push back if you only want to use the target directory.

Reply:
hmmm ... can we just expose one option like `--test-output-dir` and specify a base directory for these test output? and then change this line to something like `open(os.path.join(custom_output_dir, "%s.log" %s output_prefix), "wb+")`? I think we don't necessarily have to make it related to `target` directory, hive metastore, etc.

Reply:
I think this comes back to my original question, what is the expectation of retaining the test output? Shouldn't we retain all the files written by Spark as well? 

Independent whether we clobber the target and log directory together, the other point is, from an ergonomics perspective, is it better or worse to use a two argument option for something that is actually only one?

One alternative implementation would be to simply pass the `delete=opts.keep_test_output` flag to the `NamedTemporaryFile` and be done with it. Then we don't change any of the existing logic today. The downside is that the output will be stored in the OS's temporary locations.

Reply:
One option is to change `env["TMPDIR"] = tmp_dir` if `TMPDIR` isn't set. So users can run the script with `TMPDIR=custom_directory ./run-tests.py ....`. In that case, everything will be stored under the custom directory.

Technically test output and the files being generated by Spark are slightly different because here we capture the stdout/stderr from the top level process, and Spark runs inside the process individually.

So, setting `delete=opts.keep_test_output` would only keep the stdout/stderr from the top level process but still remove the temp files written by Spark.

Reply:
We talked about this in a chat and discussed several alternatives. The outcome of the discussion was:

1. The default location for the temporary file should be inside the temp dir that is already created.
2. The argument is a flag that toggles deletion of the directory and the tempfile.